connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -544,6 +555,95 @@ public int generation() {
         return generation;
     }
 
+    @Override
+    public synchronized void getConfigDef(final String connType, final Callback<ConfigDefInfo> callback) {
+        log.trace("Submitting get connector configuration definitions request {}", connType);
+
+        addRequest(
+            new Callable<Void>() {
+                @Override
+                public Void call() throws Exception {
+                    ConfigDef configDef = getConfig(connType);
+                    Collection<ConfigDef.ConfigKey> configKeys = configDef.configKeys().values();
+                    List<ConfigKeyInfo> configKeyInfos = new LinkedList<>();
+                    for (ConfigDef.ConfigKey configKey: configKeys) {
+                        configKeyInfos.add(convertConfigKey(configKey));
+                    }
+                    ConfigDefInfo configDefInfo = new ConfigDefInfo(connType, configKeyInfos);
+                    callback.onCompletion(null, configDefInfo);
+                    return null;
+                }
+            },
+            forwardErrorCallback(callback)
+        );
+    }
+
+    @Override
+    public synchronized void validateConfigs(final String connType, final Map<String, String> connectorConfig, final Callback<ConfigInfos> callback) {
+        log.trace("Submitting validate configuration request {}", connType);
+
+        addRequest(
+            new Callable<Void>() {
+                @Override
+                public Void call() throws Exception {
+                    ConnectorConfig connConfig = new ConnectorConfig(connectorConfig);
+                    String connName = connConfig.getString(ConnectorConfig.NAME_CONFIG);
+                    ConfigDef configDef = getConfig(connType);
+                    Connector connector = getConnector(connType, connName);
+                    List<Config> configs = connector.validate(configDef, connectorConfig);
+                    List<ConfigInfo> configInfoList = new LinkedList<>();
+                    int errorCount = 0;
+                    for (Config config : configs) {
+                        ConfigInfo configInfo = new ConfigInfo(config.getName(), config
+                            .getValue(), config.getRecommendedValues(), config.getErrorMessages(), config.isVisible());
+                        configInfoList.add(configInfo);
+                        errorCount += config.getErrorMessages().size();
+                    }
+                    ConfigInfos
+                        configInfos = new ConfigInfos(connName, errorCount, configInfoList);
+                    callback.onCompletion(null, configInfos);
+                    return null;
+                }
+            },
+            forwardErrorCallback(callback)
+        );
+
+    }
+
+    private ConfigDef getConfig(String connType) {
+        if (configs.containsKey(connType)) {
```

Comment:
I think now I see where there is shared state. Can we just make the `configs` object threadsafe on their own? If we replace one unnecessarily by using a ConcurrentHashMap here or separately synchronize, it doesn't seem like it should matter.